### ISSUE TITLE
Add metadata which says whether or not the type of authentication was persistent

### ISSUE THREAD

**Issue by rynowak** [aspnet/Security#1172]
For **Antiforgery by default** in MVC - we need a way to know that the type of authentication used was a persistent auth.

This should be implemented such that the non-persistent auth schemes (bearer) make a declaration that they **are not persistent**. The antiforgery infrastructure in MVC will look for confirmation that the scheme is **not persistent**.

**Comment by Tratcher** [comment]
What are you going to do with the information?

What auth types qualify as persistent? Is it just cookies?

**Comment by Tratcher** [comment]
Offline notes: Only Bearer qualifies as non-persistent, so we can mark that on explicitly as opting out of anti-forgery. This metadata can be carried in the AuthProperties.

@HaoK The Antiforgery filter has some trouble retrieving the AuthProperties. The AuthZ filter does its work and then sets HttpContext.User, but it does not persist the AuthProperties anywhere. You'll want to look at this when you're re-working the AuthZ filter.

**Comment by brockallen** [comment]
What makes an authentication middleware persistent or not?

**Comment by Tratcher** [comment]
It's persistent If the browser will send it on your behalf. E.g. Cookies, Windows Auth, and sometimes even Basic. These need to be protected against misuse by injected scripts, etc..

**Comment by brockallen** [comment]
Ah, ok. Persistent seemed like an odd term for that. Usually I call that an implicit credential. As always, naming is a challenge. Thx for the insight.

**Comment by Eilon** [comment]
@rynowak - putting this on your plate.

**Comment by rynowak** [comment]
Discussed this with @HaoK - for now what we're talking about is adding this metadata to https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationScheme.cs

We were talking about adding another property, but I think I like better the idea of having a property bag. @HaoK @Tratcher @davidfowl thoughts?

**Comment by HaoK** [comment]
There used to be a property bag on the Scheme called Settings, so I'm not opposed to adding it back.

Yeah we'll need to invent a way to round trip the scheme, its usually the User.Identity.AuthenticationType but that's more of a convention than any requirement.

**Comment by rynowak** [comment]
> Yeah we'll need to invent a way to round trip the scheme,

Alright, I'm going to schedule some time tomorrow where we can review this part since it sounds like I'll need to implement this

**Comment by rynowak** [comment]
Digging a little deeper into this, it looks like the place to do this is in your new PolicyEvaluator code. This is the one that actually is invoking authN policies and has both the schemes and the identities that are produced. https://github.com/aspnet/Security/pull/1193/files#diff-2626d1375c0611c39e43c0391a927bfaR41

**Comment by HaoK** [comment]
Yeah, so the policy evaluator definitely can check if all schemes are not persistent, does antiforgery run after the authorize filter?

**Comment by rynowak** [comment]
> Yeah, so the policy evaluator definitely can check if all schemes are not persistent

I guess I was imagining that the evaluator would attach the schemes to the principal, and we could check if the are persistent or not in the antiforgery filter.

> does antiforgery run after the authorize filter?

Yes, it already has to because the token is only valid for the identities we used to generate it in the first place.

**Comment by HaoK** [comment]
Ok that makes things a lot easier if the antiforgery filter runs after the authorize filter, I'm assuming there's a way for us to pass information from the authorize filter to the antiforgery, so then its just a matter of having the new policy evaluator figure out whether its authentication result is persistent.

Soo looking at https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationProperties.cs#L47

Maybe we can just use that, I'm not sure how this is used today, but if the semantics line up where its always true for cookies, and always false for bearer, this shouldn't be too bad.

**Comment by HaoK** [comment]
Altho there is the code path where context.User has already been set and the policy is relying on that ambient user.  Normally that means the DefaultAuthenticateScheme set the user, but I don't think we can rely on that, I'm not exactly sure what the soln is there, we'll have to figure that out tomorrow...

**Comment by rynowak** [comment]
After discussing more, not doing anything here for preview2

**Comment by Eilon** [comment]
We will reconsider this in 2.1.0.

**Comment by Eilon** [comment]
What are our thoughts on this for 2.1?

cc @rynowak @DamianEdwards @danroth27 @davidfowl

**Comment by rynowak** [comment]
Antiforgery by default was cut for 2.1 so I dunno

**Comment by Eilon** [comment]
@rynowak / @glennc  - do you feel this would be useful in 2.2 for the work we're doing for making APIs great?
